dali/kernels/imgproc/color_manipulation/hsv_cpu_test.cc

```diff
@@ -0,0 +1,184 @@
+using TheKernel = HsvCpu<typename GtestTypeParam::Out, typename GtestTypeParam::In>;
+
+}  // namespace
+
+
+template<class InputOutputTypes>
+class HsvCpuTest : public ::testing::Test {
+ protected:
+    HsvCpuTest() {
+        input_.resize(dali::volume(shape_));
+        ref_output_.resize(dali::volume(shape_));
+    }
+
+
+    void SetUp() final {
+        std::mt19937_64 rng;
+        UniformRandomFill(input_, rng, 0., 10.);
+        calc_output<typename InputOutputTypes::Out>();
+        ref_out_tv_ = make_tensor_cpu(ref_output_.data(), shape_);
+    }
+
+
+    std::vector<typename InputOutputTypes::In> input_;
+    std::vector<typename InputOutputTypes::Out> ref_output_;
+    OutTensorCPU<typename InputOutputTypes::Out, ndims> ref_out_tv_;
+    TensorShape<3> shape_ = {23, 45, ndims};
+    float hue_ = 2.8f;
+    float saturation_ = 1.9f;
+    float value_ = 0.3f;
+
+
+    template<typename OutputType>
+    std::enable_if_t<std::is_integral<OutputType>::value> calc_output() {
+        for (size_t i = 0; i < input_.size(); i += ndims) {
+            ref_output_[i] = std::round(input_[i] + hue_);
+            ref_output_[i + 1] = std::round(input_[i + 1] * saturation_);
+            ref_output_[i + 2] = std::round(input_[i + 2] * value_);
+        }
+    }
+
+
+    template<typename OutputType>
+    std::enable_if_t<!std::is_integral<OutputType>::value> calc_output() {
+        for (size_t i = 0; i < input_.size(); i += ndims) {
+            ref_output_[i] = input_[i + 0] + hue_;
+            ref_output_[i + 1] = input_[i + 1] * saturation_;
+            ref_output_[i + 2] = input_[i + 2] * value_;
+        }
+    }
+};
+
+
+using TestTypes = std::tuple<uint8_t, int8_t, uint16_t, int16_t, int32_t, float>;
+INPUT_OUTPUT_TYPED_TEST_SUITE(HsvCpuTest, TestTypes);
+
+
+TYPED_TEST(HsvCpuTest, check_kernel) {
+    TheKernel<TypeParam> kernel;
+    check_kernel<decltype(kernel)>();
+}
+
+
+TYPED_TEST(HsvCpuTest, SetupTestAndCheckKernel) {
+    TheKernel<TypeParam> kernel;
+    KernelContext ctx;
+    InTensorCPU<typename TypeParam::In, ndims> in(this->input_.data(), this->shape_);
+    auto reqs = kernel.Setup(ctx, in, this->hue_, this->saturation_, this->value_);
+    auto sh = reqs.output_shapes[0][0];
+    ASSERT_EQ(this->shape_, sh);
+}
+
+
+TYPED_TEST(HsvCpuTest, RunTest) {
+    TheKernel<TypeParam> kernel;
+    KernelContext ctx;
+    InTensorCPU<typename TypeParam::In, ndims> in(this->input_.data(), this->shape_);
+    auto reqs = kernel.Setup(ctx, in, this->hue_, this->saturation_, this->value_);
+    auto out_shape = reqs.output_shapes[0][0];
+    vector<typename TypeParam::Out> output;
+    output.resize(dali::volume(out_shape));
+    OutTensorCPU<typename TypeParam::Out, ndims> out(output.data(),
+                                                     out_shape.template to_static<ndims>());
+
+    kernel.Run(ctx, out, in, this->hue_, this->saturation_, this->value_);
+    for (int i = 0; i < out.num_elements(); i++) {
```

Comment:
you could use the test kernel util `Check`

Reply:
`Check` code consists of few overloads of this function in ~200 LOC, without any documentation. Frankly, it's way quicker for me to write it on my own and then I know precisely what actually happens than to figure it out by reading these functions.
On top of that, whoever reads my piece of code would also need to figure out what precisely happens in `Check`. So I'd rather write these 3 lines of code than replace it by function, which has 100...